src/installer.ts

```diff
@@ -517,6 +556,37 @@ export function parseNodeVersionFile(contents: string): string {
   return nodeVersion as string;
 }
 
-function isLatestSyntax(versionSpec): boolean {
+export function isLatestSyntax(versionSpec): boolean {
   return ['current', 'latest', 'node'].includes(versionSpec);
 }
+
+export function evaluateCanaryMatcher(
+  versionSpec: string
+): (potential: string) => boolean {
+  const [raw, prerelease] = versionSpec.split(/-(.*)/s);
+  const isValidVersion = semver.valid(raw);
+  const rawVersion = isValidVersion ? raw : semver.coerce(raw)?.version;
+  if (rawVersion) {
+    if (prerelease === 'v8-canary') {
+      // it means versionSpec does not have timestamp
+      const range = semver.validRange(`^${rawVersion}`);
+      return (potential: string) =>
+        semver.satisfies(
+          // TODO: check latest?
+          potential.replace('-v8-canary', '+v8-canary.'),
+          range
+        );
+    } else {
+      // see https://github.com/actions/setup-node/blob/00e1b6691b40cce14b5078cb411dd1ec7dab07f7/__tests__/verify-node.sh#L10
+      // there must be exact match
+      const range = `${rawVersion}-${prerelease}`;
+      return (potential: string) =>
+        semver.satisfies(
+          // TODO: check latest?
+          potential,
+          range
+        );
+    }
+  }
+  return () => false;
+}
```

Comment:
Re: our Slack discussion.  Using polymorphism like this is a good solution to avoid having to put `if` statements all over the code for the different cases.  You can invoke the Factory Pattern directly to make it a bit more obvious to a reader.  Also, I would bring the other matcher for non-canary versions into this factory function.

```suggestion
interface VersionMatcher {
  (potential: string): boolean
}

export function versionMatcherFactory(versionSpec: string): VersionMatcher {
  if (!isVersionCanary(versionSpec)) {
    return potential => semver.satisfies(potential, versionSpec);
  }

  const [raw, prerelease] = versionSpec.split(/-(.*)/s);
  const rawVersion = semver.valid(raw) ? raw : semver.coerce(raw)?.version;
  if (rawVersion) {
    if (prerelease === 'v8-canary') {
      // it means versionSpec does not have timestamp
      const range = semver.validRange(`^${rawVersion}`);
      return (potential: string) =>
        semver.satisfies(
          // TODO: check latest?
          potential.replace('-v8-canary', '+v8-canary.'),
          range
        );
    } else {
      // see https://github.com/actions/setup-node/blob/00e1b6691b40cce14b5078cb411dd1ec7dab07f7/__tests__/verify-node.sh#L10
      // there must be exact match
      const range = `${rawVersion}-${prerelease}`;
      return (potential: string) =>
        semver.satisfies(
          // TODO: check latest?
          potential,
          range
        );
    }
  } else {
    // TODO Is this case ever intended?
    return () => false;
  }
}
```

Reply:
resolved in the current branch